### PR TITLE
Add a Gibbs sampler for the negative-binomial dispersion parameter

### PR DESCRIPTION
**Pull request by xjing76**
This PR will be adding Gibbs steps that would allow sampling for dispersion terms of Negative Binomial. 

This PR will follow the compound poisson method mentioned [here](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4180062/)

So far we roughy added the Gibbs steps for  
1. r -dispersion terms in N(r, p)
2. h - Inverse of scale parameter for gamma distribution of r
3. L_i - The number terms of L/N of a Poisson distributed variable in Compound Poisson distribution

The most tricky part so far comes form the sampling of L_i as it is a discrete sampling from a given probability vector where the probability vector comes from a pre-calculated matrix with some additional computation. 

Next step is to add `match` functions for the NB graph with dispersion terms.

### CODE REVIEW

**Comment by zoj613** [comment]
@xjing76 You might want to consider setting up `pre-commit` locally using the config file in this repo, Tests for the style check are currently failing.

**Comment by rlouf** [comment]
Great! Following what @zoj613 said: could you install `pre-commit` and make sure that /every commit/ at least passes code checks, and to not start on a new commit until the current one passes the tests as well?

**Comment by xjing76** [comment]
I think I set up the main steps for the alpha gibbs sampler steps. 

1. However, I think I need a more simple way to calculate the pre-calculated matrix F  as the size of it is determined by the maximum of observations, and so the size of it can get quite large. 

2. I am still looking for ways to make this`srng.choice(range(N), size=1, replace=True, p=R_r(F, r, i)) for i in y` work as p seems cannot be assigned as a rv.

**Comment by xjing76** [comment]
> Great! For (2) could you provide a minimal example that illustrates the problem you're having so I can have a look?

I think the problem itself is not what I thought it was initially. But the problem still persist. 

1.  Initially i set up the sampling step as 

But it seems that I am not able to iterate through y (with a `can't determine dimension` error)
li = [srng.choice(at.arange(N), size=1, replace=True, p=R_r(F, r, i)) for i in y]

But base on the paper equation (23-25) we would need to sample `l_i` seperately for each new `r` and each `y_i`

2. Just to test out the technical feasibility of the set up, I did

li = srng.choice(at.arange(N), size=N, replace=True, p=R_r(F, r, 10)) 

This would not work as well as some step results p vector to be a vector full of `nan` 

Any suggestions how should I debug this?

**Comment by rlouf** [comment]
Do you have a setup that allows you to work with a debugger such as [pdb](https://docs.python.org/3/library/pdb.html)? You should then be able to go up the stack trace when said line fails and investigate where these `NaN`s come from.

**Comment by xjing76** [comment]
> Do you have a setup that allows you to work with a debugger such as [pdb](https://docs.python.org/3/library/pdb.html)? You should then be able to go up the stack trace when said line fails and investigate where these `NaN`s come from.

I do have `--pdb` set up but the results are coming from graph op and there are more arguments than what we supply. I will take a further look into it. Thanks!

**Comment by xjing76** [comment]
1 On making `F` and `R_r` matrix numerically stable. However I didn't use the stirling method that was mentioned in [this paper](https://scholar.google.com/scholar?as_q=&as_epq=Augment-and-conquer%20negative%20binomial%20processes&as_occt=title&as_sauthors=Mingyuan+Zhou&as_ylo=2012&as_yhi=2012&as_sdt=1.&as_sdtp=on&as_sdtf=&as_sdts=22&) here (lemma 2.1) as Stirling number can get overflow pretty easily at 100. I just added as switch state as in the paper that when r > 1 we would use the unweighted `F` instead of weighted matrix `R_r`

2. Since `ChoiceRV` in Aesara only take a vector as a p, I am coming up with a `multichoiceRV` here (which is not working yet) to temporarily solve this issue. But I am hoping that we might eventually adding a `choiceRV` in aesara that would be able to handle such problem.

File: aemcmc/gibbs.py [hunk 1]
```diff
@@ -227,6 +229,141 @@ def nbinom_sigmoid_dot_match(
...
+        return F[0, :]
+    F = at.as_tensor_variable(F)
+    r_factor = at.exp(at.log(r_) * at.arange(1, n + 1))
+    r_factor = at.switch(at.lt(r_factor, 1), r_factor, 1)
+    dinominator = F[row - 1, :] @ r_factor
+    nominator = F[row - 1, :] * r_factor
+    return nominator / dinominator
+
+
+class MultiChoiceRV(ChoiceRV):
```
Comment (rlouf) [inline aemcmc/gibbs.py:269]: Could you move this to `dist.py`?

Comment (xjing76) [reply aemcmc/gibbs.py:269]: Oh. of course.

**Comment by xjing76** [comment]
So far, I think the engineering for Multichoice RV are done. The shape and unit test are passing without much issue. However, it seems like the sampling step itself is converging. Looking at the samples for R, it keep getting larger and larger while the true dispersion term is only 1. It would be the next step for me.

**Comment by xjing76** [comment]
By only looking at `test_horseshoe_nbinom` the betas sampled in those set up are not converging very well either. So I set up a [separate branch](https://github.com/xjing76/aemcmc/tree/fixed_beta) with only given betas and allow only sampling for the dispersion parts. 

However, it seems like the dispersion portion is still converging to a very large number than what we preset to (1). I attempt to fully covert the code to the implementation of [matlab version](https://mingyuanzhou.github.io/Softwares/LGNB_Regression_v0.zip) yet i think still need some more time to figure out the model as the `phi` portion used in matlab is intertwined with the sampling of both beta and the dispersion `r`


https://github.com/xjing76/aemcmc/tree/fixed_beta

File: tests/test_dists.py [hunk 2]
```diff
@@ -31,6 +32,25 @@ def test_polyagamma():
...
+    
+    rng2 = at.random.RandomStream(12345)
+    res2 = rng2.gen(multichoice, np.array(10, dtype = np.int64), 1, True, p_val)
+    # add assersion fpr results
+    assert np.alltrue(res1.eval() == res2.eval())
+    assert res1.eval().dtype == "int64"
+    
+
+    res3 = rng1.gen(multichoice, np.arange(10), 2, True, p_val)
+    # breakpoint()
```
Comment (brandonwillard) [inline tests/test_dists.py:50]:
```suggestion
```

File: tests/test_dists.py [hunk 3]
```diff
@@ -31,6 +32,25 @@ def test_polyagamma():
...
+def test_multichoice():
+
+    rng1 = at.random.RandomStream(12345)
+    p_val = np.stack([np.repeat(0.1, 10)] * 3, axis=0)
+    res1 = rng1.gen(multichoice, np.array(10, dtype = np.int64), 1, True, p_val)
+    
+    rng2 = at.random.RandomStream(12345)
+    res2 = rng2.gen(multichoice, np.array(10, dtype = np.int64), 1, True, p_val)
+    # add assersion fpr results
+    assert np.alltrue(res1.eval() == res2.eval())
```
Comment (brandonwillard) [inline tests/test_dists.py:45]: The graphs `res1` and `res2` are literally the same, no?  It's not clear what this `assert` is supposed to be testing, and, as the comment says, we absolutely need to confirm that this is working before considering the results of a model that uses this sampler.

File: tests/test_gibbs.py [hunk 4]
```diff
@@ -132,6 +136,60 @@ def test_horseshoe_nbinom(srng):
     assert np.all(lmbda > 0)
 
 
+def test_R_r(srng):
+    F = F_matrix_construct(50)
+    r_ = srng.gamma(1, 1)
+    R_r(F, r_, 4)
```
Comment (brandonwillard) [inline tests/test_gibbs.py:142]: This looks like an unfinished test. We definitely need a test for these components (i.e. everything involved in `sampling_for_li`, and that function itself), though.

**Review by brandonwillard — CHANGES_REQUESTED** [review]
> I think it may not be enough just to have a test on the shape of the samples but also on the values of those sample results.

Yes, a solid set of unit tests for `sampling_for_li` are needed.

> By only looking at `test_horseshoe_nbinom` the betas sampled in those set up are not converging very well either. So I set up a [separate branch](https://github.com/xjing76/aemcmc/tree/fixed_beta) with only given betas and allow only sampling for the dispersion parts.
> 
> However, it seems like the dispersion portion is still converging to a very large number than what we preset to (1). I attempt to fully covert the code to the implementation of [matlab version](https://mingyuanzhou.github.io/Softwares/LGNB_Regression_v0.zip) yet i think still need some more time to figure out the model as the `phi` portion used in matlab is intertwined with the sampling of both beta and the dispersion `r`

To which values/output are you referring?  Ones from `test_horseshoe_nbinom_w_dispersion`?  I'm not able to get that test to run locally due to unresolved shape issues, and there aren't any results in CI with which to compare due to the `pre-commit` errors.

My guess is that the `MultiChoiceRV` implementation isn't finished.  There are definitely some important test cases missing, so the next step is to add some more thorough tests.

When we're confident that `MultiChoiceRV` is working as expected, we'll need to do the same for the distribution implemented by `sampling_for_li`.

**Comment by brandonwillard** [comment]
> I attempt to fully covert the code to the implementation of [matlab version](https://mingyuanzhou.github.io/Softwares/LGNB_Regression_v0.zip) yet i think still need some more time to figure out the model as the `phi` portion used in matlab is intertwined with the sampling of both beta and the dispersion `r`

That implementation references M. Zhou and L. Carin, "Negative Binomial Process Count and Mixture Modeling,"  arXiv:1209.3442, Sept. 2012 and M. Zhou and L. Carin, "Augment-and-Conquer Negative Binomial Processes," in NIPS 2012.  The former uses the CRTP distribution formulation for `L_i`, instead of the computations you're using from Zhou M, Li L, Dunson D, Carin L. "Lognormal and Gamma Mixed Negative Binomial Regression", and the latter states that the CRTP distribution is a sum of CRT variates, and that CRT variates can be drawn as a sum of Bernoullis.

The CRTP approach looks much simpler, as the reference implementation demonstrates, so we should go with that one.

**Comment by xjing76** [comment]
Here I finally updated this branch with similar set up as with the matlab implementation. 

1 I set up a CRT_sum and the sampling step for dispersion term "r" almost exactly like that of in the matlab implementation. However, I kept the sampling step for `beta` the same way as it was before. 

2 Here is still quite some issues with the current set up. 
   - the steps do quite badly with any 0 impressions. I had to tune the toy data parameter to make sure all y are above zero. 
   - The sampling of `r` are making huge jumps between very large value e6 and very small values e-5, which is also one of the reason why the polygamma sampling is problematic  with impressions have zeros in them, as the `h` parameter, composed of `y+ r`. 
   - Technical wise the sampling step is doing okay with good speed and should be bug free however, the convergence is till not very ideal

**Comment by brandonwillard** [comment]
Don't forget to resolve the open conversations once they've been addressed.

File: aemcmc/gibbs.py [hunk 5]
```diff
@@ -227,6 +229,54 @@ def nbinom_sigmoid_dot_match(
...
+    y: TensorVariable,
+    xb: TensorVariable,
+    T: int,
+) -> Tuple[TensorVariable, TensorVariable]:
+    PsiSig = 1.0 / (phi + w)
+    PsiMu = PsiSig @ ((y - r) / 2 + phi * (xb))
+    rng = np.random.RandomState(2313)
+    Psi = PsiMu + rng.randint(T) * at.sqrt(PsiSig)
+    psi_beta = (Psi - xb) / Psi
+    phi = srng.gamma(1e-2 + T / 2, 1 / (1e-2 + at.sum(psi_beta * psi_beta) / 2), size=1)
```
Comment (brandonwillard) [inline aemcmc/gibbs.py:260]: These references to a $\psi$ term lead me to believe that these steps aren't specifically for the $r$ parameter, but other parts of the model that are already covered by the existing PG-based negative-binomial sampler.

That could very easily explain any issues you've seen with this implementation.

File: aemcmc/gibbs.py [hunk 6]
```diff
@@ -364,6 +423,129 @@ def nbinom_horseshoe_step(
...
+            srng, beta_new, 1.0, lmbda_inv, tau_inv
+        )
+
+        r_new, phi_new = dispersion_term_step(srng, phi, r, w, y, xb, T)
+
+        return beta_new, 1.0 / lmbda_inv_new, 1.0 / tau_inv_new, r_new, phi_new
+
+    X, beta_rv, lmbda_rv, tau_rv, r_rv = nbinom_horseshoe_match_with_dispersion(Y_rv)
+
+    phi_rv = srng.gamma(1, 1, size=1)
```
Comment (brandonwillard) [inline aemcmc/gibbs.py:536]: What are these `phi*` terms?  Also, a sampling step for $r$ should only require the previous sample value of $r$ and the $\pi_i$ that make up $y_i \sim \operatorname{NB}\left(r, \pi_i\right)$.

**Review by brandonwillard — CHANGES_REQUESTED** [review]
In general, it looks like you've adapted the reference Matlab code a bit too literally.  Extra terms have been introduced that appear to overlap with existing terms already handled by `horseshoe_step`.

It would really help to write clear docstrings for these new functions so that we can more easily see the mathematical assumptions.

File: tests/test_gibbs.py [hunk 7]
```diff
@@ -132,6 +133,60 @@ def test_horseshoe_nbinom(srng):
...
+    sample_fn = aesara.function((num_samples,), outputs, updates=updates)
+
+    sample_num = 2000
+    beta, lmbda, tau, r = sample_fn(sample_num)
+
+    y_post = []
+    for i in range(sample_num):
+        y_post.append(srng.nbinom(r[i], at.sigmoid(-(X.dot(beta[i])))).eval())
+
+    np.array(y_post).mean(0)
```
Comment (brandonwillard) [inline tests/test_gibbs.py:178]: This looks like an unfinished line.

File: aemcmc/gibbs.py [hunk 8]
```diff
@@ -250,6 +269,54 @@ def nbinom_horseshoe_match(
...
+
+
+def r_step(
+    srng: RandomStream,
+    r: TensorVariable,
+    p: TensorVariable,
+    a_0: TensorVariable,
+    h: TensorVariable,
+    y: TensorVariable,
+) -> Tuple[TensorVariable, TensorVariable, TensorVariable]:
```
Comment (brandonwillard) [inline aemcmc/gibbs.py:304]: This still needs a docstring.

**Review by brandonwillard — CHANGES_REQUESTED** [review]
Don't forget to resolve open conversations (that have been resolved via code changes).

**Review by brandonwillard — CHANGES_REQUESTED** [review]
The rename of `h*` to `r*` should be in a separate commit.

File: aemcmc/gibbs.py [hunk 9]
```diff
@@ -250,6 +269,54 @@ def nbinom_horseshoe_match(
...
+    eta = X @ beta_rv
+    p = at.sigmoid(-eta)
+    r = srng.gamma(a_0, h, size=1)
+    Y_rv = srng.nbinom(r, p)
+    return Y_rv
+
+
+def crtsum_rand(srng, x, r):
+    rnd_temp = r / (r + at.arange(at.max(x)))
+    Lsum = 0
```
Comment (brandonwillard) [inline aemcmc/gibbs.py:288]:
```suggestion
    L_sum = 0
```

File: aemcmc/gibbs.py [hunk 10]
```diff
@@ -364,6 +430,129 @@ def nbinom_horseshoe_step(
...
+        tau_inv = 1.0 / tau
+        beta_new = update_beta(srng, w, lmbda_inv * tau_inv, X, z)
+
+        lmbda_inv_new, tau_inv_new = horseshoe_step(
+            srng, beta_new, 1.0, lmbda_inv, tau_inv
+        )
+        eta = X @ beta_new
+        p = at.sigmoid(-eta)
+        h = srng.gamma(1e-2 + 1 * 1e-2, 1 / (1e-2 + r))
+        r_new = r_step(srng, r, p, a_0, h, y)
```
Comment (brandonwillard) [inline aemcmc/gibbs.py:537]: This `srng.gamma` is a fixed (hyper)prior assumption that we can't have baked into this sampler.  Instead, we want to let the user define their own models/(hyper)priors&mdash;and corresponding sample steps.
```suggestion
        r_new = r_step(srng, r, p, a_0, h, y)
```

File: aemcmc/gibbs.py [hunk 11]
```diff
@@ -364,6 +430,129 @@ def nbinom_horseshoe_step(
...
+
+    def nbinom_horseshoe_step(
+        beta: TensorVariable,
+        lmbda: TensorVariable,
+        tau: TensorVariable,
+        r: TensorVariable,
+        y: TensorVariable,
+        X: TensorVariable,
+        a_0: TensorVariable,
+    ) -> Tuple[TensorVariable, TensorVariable, TensorVariable, TensorVariable]:
```
Comment (brandonwillard) [inline aemcmc/gibbs.py:500]:
```suggestion
        a_0: TensorVariable,
        h: TensorVariable,
    ) -> Tuple[TensorVariable, TensorVariable, TensorVariable, TensorVariable]:
```

File: aemcmc/gibbs.py [hunk 12]
```diff
@@ -364,6 +430,129 @@ def nbinom_horseshoe_step(
...
+        eta = X @ beta_new
+        p = at.sigmoid(-eta)
+        h = srng.gamma(1e-2 + 1 * 1e-2, 1 / (1e-2 + r))
+        r_new = r_step(srng, r, p, a_0, h, y)
+
+        return beta_new, 1.0 / lmbda_inv_new, 1.0 / tau_inv_new, r_new
+
+    X, beta_rv, lmbda_rv, tau_rv, r_rv = nbinom_horseshoe_match_with_dispersion(Y_rv)
+
+    a_0, _ = gamma_prior_match(r_rv)
```
Comment (brandonwillard) [inline aemcmc/gibbs.py:543]:
```suggestion
    a_0, h = gamma_prior_match(r_rv)
```

File: aemcmc/gibbs.py [hunk 13]
```diff
@@ -364,6 +430,129 @@ def nbinom_horseshoe_step(
...
+
+    outputs, updates = aesara.scan(
+        nbinom_horseshoe_step,
+        outputs_info=[beta_rv, lmbda_rv, tau_rv, r_rv],
+        non_sequences=[y, X, a_0],
```
Comment (brandonwillard) [inline aemcmc/gibbs.py:548]:
```suggestion
        non_sequences=[y, X, a_0, h],
```

**Review by brandonwillard — CHANGES_REQUESTED** [review]

**Comment by brandonwillard** [comment]
The current failure is due the problem addressed in https://github.com/aesara-devs/aesara/pull/1003.

**Review by brandonwillard — APPROVED** [review]

**Pull request merged**